### PR TITLE
✨ FixedMathLib: expWadDown()

### PR DESCRIPTION
**Pull request by recmo**
## Description

Describe the changes made in your pull request here.

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [ ] Ran `dapp snapshot`?
- [ ] Ran `npm run lint`?
- [x] Ran `forge test`?
- [ ] Ran `dapp test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

**Comment by recmo** [comment]
Ping @transmissions11

**Comment by transmissions11** [comment]
👀👀👀👀👀👀👀👀👀

**Comment by transmissions11** [comment]
amazing already

Screen Shot 2022-03-21 at 11 22 47 PM

**Comment by recmo** [comment]
Hmm... that should have been a ~10 gas reduction.

**Comment by transmissions11** [comment]
well if we inline everything into q should save?

**Comment by recmo** [comment]
Can't inline the variables `z` and `w` because they both get used twice. That's the scheme basically, you need more variables and additions, but in return you have fewer multiplies. Multiplies are more expensive, so it may be worthwhile.

**Comment by transmissions11** [comment]
yup... thought maybe this would save some but looks like the optimizer was smart enough in the first place

Screen Shot 2022-03-21 at 11 47 08 PM

**Comment by recmo** [comment]
We can always go back to the original formula for `q`, but somehow I feel that removing a multiply should save us something :D. 

I have a similar scheme for `p` that also removes a mul:

```
int256 y = x      +     1346386616545796478920950773328;
y = (y * x >> 96) +    57155421227552351082224309758442;
int256 p = y + x  -    94201549194550492254356042504812;
p = (p * y >> 96) + 28719021644029726153956944680412240;
```

but that one somehow ends up costing 1 gas more than the plain formula in my local tests.

**Comment by transmissions11** [comment]
the scheme for p saves gas in my tests!

Screen Shot 2022-03-21 at 11 51 56 PM

**Comment by transmissions11** [comment]
want me to commit? seems like the scheme for p and the original formula for q is the fastest

**Comment by recmo** [comment]
lol! In my repo it's exact opposite! Yeah, go a head an use whatever works best. But I wonder what's going on though.

File: src/test/FixedPointMathLib.t.sol [hunk 1]
```diff
@@ -6,12 +6,54 @@ import {DSTestPlus} from "./utils/DSTestPlus.sol";
...
+        assertEq(FixedPointMathLib.expWadDown(-2e18), 135335283236612691);
+        assertEq(FixedPointMathLib.expWadDown(-1e18), 367879441171442321);
+
+        assertEq(FixedPointMathLib.expWadDown(-0.5e18), 606530659712633423);
+        assertEq(FixedPointMathLib.expWadDown(-0.3e18), 740818220681717866);
+
+        assertEq(
+            FixedPointMathLib.expWadDown(0),
+            999999999999999999
+            // True value: 1000000000000000000 (off by 1 wei)
```
Comment (recmo) [inline src/test/FixedPointMathLib.t.sol:22]: This actually annoys me quite a bit.

File: src/utils/FixedPointMathLib.sol [hunk 2]
```diff
@@ -27,6 +27,64 @@ library FixedPointMathLib {
...
+
+            // k is in the range [-61, 195].
+
+            // Evaluate using a (6, 7)-term rational approximation.
+            // p is made monic, we'll multiply by a scale factor later.
+            int256 y = x + 1346386616545796478920950773328;
+            y = ((y * x) >> 96) + 57155421227552351082224309758442;
+            int256 p = y + x - 94201549194550492254356042504812;
+            p = ((p * y) >> 96) + 28719021644029726153956944680412240;
+            p = p * x + (4385272521454847904632057985693276 << 96);
```
Comment (transmissions11) [inline src/utils/FixedPointMathLib.sol:59]: should we add a comment about using knuth's method here?

File: src/test/FixedPointMathLib.t.sol (hunk 1 above)
Comment (transmissions11) [reply src/test/FixedPointMathLib.t.sol:22]: lmk adding to the constant term works without impacting the other vars

another way thats only a couple extra opcodes would be doing 

```
           assembly {
            	r := add(r, iszero(x))
            }
```

right at the end

**Comment by transmissions11** [comment]
btw how hard would it be to implement a rounding up version of this? looks like this impl truncates (which is great but i'd like to have both a floor and ceil version)

File: src/utils/FixedPointMathLib.sol (hunk 2 above)
Comment (recmo) [reply src/utils/FixedPointMathLib.sol:59]: Sure, it's the first method from page 491. He mentions a second one which has one fewer addition, but leaves the details as an exercise. I haven't bother to do that exercise yet :)

**Comment by recmo** [comment]
> btw how hard would it be to implement a rounding up version of this? looks like this impl truncates (which is great but i'd like to have both a floor and ceil version)

This should work (assuming we fix the exp(0) = 1 issue):

```solidity
function expCeil(int256 x) returns (int256) {
   if (x == 0) return 1e18; // ceil(exp(0)) = floor(exp(0)) = 1
   return exp(x) + 1;
}
```

**Comment by recmo** [comment]
But I don't think it is guaranteed to return the truncated / floor result. For larger input values it doesn't have the precision required to do accurate rounding.

It also doesn't make sense to increase the precision for large inputs, because the input itself is only accurate up to 1e-18.

**Comment by transmissions11** [comment]
> But I don't think it is guaranteed to return the truncated / floor result. For larger input values it doesn't have the precision required to do accurate rounding.

oh hmm yea maybe just got lucky with the test cases i used

**Pull request merged**